src/convert.rs

```diff
@@ -0,0 +1,192 @@
+    //
+    //Second, the from and into for the same color must override the default.
+    //For example, from_rgb and into_rgb for Rgb will convert via Xyz which
+    //needs to be overridden with *input_color and *self to avoid the unnecessary
+    //converison.
+    //
+    //Third, if a direct transform exists between the colors, then it should be
+    //overridden in both the colors. For example, Luma has direct conversion to
+    //Rgb. So into_rgb and from_rgb conversion is implemented in Luma and into_luma
+    //and from_luma is overridden in Rgb to use the conversion from Luma.
+
+    ///Convert from XYZ color space
+    fn from_xyz(&Xyz<T>) -> Self;
+
+    ///Convert into xyz space
+    fn into_xyz(&self) -> Xyz<T>;
+
+    ///Convert from Yxy color space
+    fn from_yxy(inp: &Yxy<T>) -> Self {
+        Self::from_xyz(&inp.into_xyz())
+    }
+
+    ///Convert into Yxy color space
+    fn into_yxy(&self) -> Yxy<T> {
+        Yxy::from_xyz(&self.into_xyz())
+    }
+
+    ///Convert from Lab color space
+    fn from_lab(inp: &Lab<T>) -> Self {
+        Self::from_xyz(&inp.into_xyz())
+    }
+
+    ///Convert into Lab color space
+    fn into_lab(&self) -> Lab<T> {
+        Lab::from_xyz(&self.into_xyz())
+    }
+
+    ///Convert from Lch color space
+    fn from_lch(inp: &Lch<T>) -> Self {
+        Self::from_lab(&inp.into_lab())
+    }
+
+    ///Convert into Lch color space
+    fn into_lch(&self) -> Lch<T> {
+        Lch::from_lab(&self.into_lab())
+    }
+
+    ///Convert from Rgb color space
+    fn from_rgb(inp: &Rgb<T>) -> Self {
+        Self::from_xyz(&inp.into_xyz())
+    }
+
+    ///Convert into Rgb color space.
+    fn into_rgb(&self) -> Rgb<T> {
+        Rgb::from_xyz(&self.into_xyz())
+    }
+
+    ///Convert from Hsl color space
+    fn from_hsl(inp: &Hsl<T>) -> Self {
+        Self::from_rgb(&inp.into_rgb())
+    }
+
+    ///Convert into Hsl color space
+    fn into_hsl(&self) -> Hsl<T> {
+        Hsl::from_rgb(&self.into_rgb())
+    }
+
+    ///Convert from Hsv color space
+    fn from_hsv(inp: &Hsv<T>) -> Self {
+        Self::from_rgb(&inp.into_rgb())
+    }
+
+    ///Convert into Hsv color space
+    fn into_hsv(&self) -> Hsv<T> {
+        Hsv::from_rgb(&self.into_rgb())
+    }
+
+    ///Convert from Luma
+    fn from_luma(inp: &Luma<T>) -> Self {
+        Self::from_xyz(&inp.into_xyz())
+    }
+
+    ///Convert into Luma
+    fn into_luma(&self) -> Luma<T> {
+        Luma::from_xyz(&self.into_xyz())
+    }
+
+}
+
+macro_rules! impl_from_trait {
+    (($self_ty:ident, $into_fn: ident) => {$($other:ident),+}) => (
+        impl<T: Float> From<Alpha<$self_ty<T>, T>> for $self_ty<T> {
+            fn from(color: Alpha<$self_ty<T>, T>) -> $self_ty<T> {
+                color.color
+            }
+        }
+
+        impl<T: Float> From<Color<T>> for $self_ty<T> {
+            fn from(color: Color<T>) -> $self_ty<T> {
+                match color {
+                    Color::Luma(c) => c.$into_fn(),
```

Comment:
Sorry, I meant rewriting the `Color::Luma(c) => c.$into_fn(),` lines to `$(Color::$other(c) => c.$into_fn()),+`.